src/c_api.cpp

```diff
@@ -0,0 +1,841 @@
+  int is_row_major,
+  int predict_type,
+  int64_t n_used_trees,
+  double* out_result) {
+
+  Booster* ref_booster = reinterpret_cast<Booster*>(handle);
+  ref_booster->PrepareForPrediction(static_cast<int>(n_used_trees), predict_type);
+
+  auto get_row_fun = RowPairFunctionFromDenseMatric(data, nrow, ncol, data_type, is_row_major);
+  int num_class = ref_booster->NumberOfClasses();
+#pragma omp parallel for schedule(guided)
+  for (int i = 0; i < nrow; ++i) {
+    auto one_row = get_row_fun(i);
+    auto predicton_result = ref_booster->Predict(one_row);
+    for (int j = 0; j < num_class; ++j) {
+      out_result[i * num_class + j] = predicton_result[j];
+    }
+  }
+  return 0;
+}
+
+DllExport int LGBM_BoosterSaveModel(BoosterHandle handle,
+  int num_used_model,
+  const char* filename) {
+
+  Booster* ref_booster = reinterpret_cast<Booster*>(handle);
+  ref_booster->SaveModelToFile(num_used_model, filename);
+  return 0;
+}
+
+// ---- start of some help functions
+
+std::function<std::vector<double>(int row_idx)>
+RowFunctionFromDenseMatric(const void* data, int num_row, int num_col, int data_type, int is_row_major) {
+  if (data_type == dtype_float32) {
+    const float* data_ptr = reinterpret_cast<const float*>(data);
+    if (is_row_major) {
+      return [data_ptr, num_col, num_row](int row_idx) {
+        CHECK(row_idx < num_row);
+        std::vector<double> ret;
+        auto tmp_ptr = data_ptr + num_col * row_idx;
+        for (int i = 0; i < num_col; ++i) {
+          ret.push_back(static_cast<double>(*(tmp_ptr + i)));
+        }
+        return ret;
+      };
+    } else {
+      return [data_ptr, num_col, num_row](int row_idx) {
+        CHECK(row_idx < num_row);
+        std::vector<double> ret;
+        for (int i = 0; i < num_col; ++i) {
+          ret.push_back(static_cast<double>(*(data_ptr + num_row * i + row_idx)));
+        }
+        return ret;
+      };
+    }
+  } else if (data_type == dtype_float64) {
+    const double* data_ptr = reinterpret_cast<const double*>(data);
+    if (is_row_major) {
+      return [data_ptr, num_col, num_row](int row_idx) {
+        CHECK(row_idx < num_row);
+        std::vector<double> ret;
+        auto tmp_ptr = data_ptr + num_col * row_idx;
+        for (int i = 0; i < num_col; ++i) {
+          ret.push_back(static_cast<double>(*(tmp_ptr + i)));
+        }
+        return ret;
+      };
+    } else {
+      return [data_ptr, num_col, num_row](int row_idx) {
+        CHECK(row_idx < num_row);
+        std::vector<double> ret;
+        for (int i = 0; i < num_col; ++i) {
+          ret.push_back(static_cast<double>(*(data_ptr + num_row * i + row_idx)));
+        }
+        return ret;
+      };
+    }
+  } else {
+    Log::Fatal("unknown data type in RowFunctionFromDenseMatric");
+  }
+  return nullptr;
+}
+
+std::function<std::vector<std::pair<int, double>>(int row_idx)>
+RowPairFunctionFromDenseMatric(const void* data, int num_row, int num_col, int data_type, int is_row_major) {
```

Comment:
maybe there are more elegant way to combine code between +RowPairFunctionFromDenseMatric  +RowFunctionFromDenseMatric ?